src/docker/shared.rs

```diff
@@ -488,6 +489,271 @@ impl Directories {
+            container_stop(
+                &deleter.engine,
+                &deleter.name,
+                deleter.timeout,
+                &mut msg_info,
+            )
+            .ok();
+            container_rm(&deleter.engine, &deleter.name, &mut msg_info).ok();
+
+            self.deleter = None;
+        }
+    }
+}
+
+impl Drop for Container {
+    fn drop(&mut self) {
+        self.terminate();
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum ContainerState {
+    Created,
+    Running,
+    Paused,
+    Restarting,
+    Dead,
+    Exited,
+    DoesNotExist,
+}
+
+impl ContainerState {
+    pub fn new(state: &str) -> Result<Self> {
+        match state {
+            "created" => Ok(ContainerState::Created),
+            "running" => Ok(ContainerState::Running),
+            "paused" => Ok(ContainerState::Paused),
+            "restarting" => Ok(ContainerState::Restarting),
+            "dead" => Ok(ContainerState::Dead),
+            "exited" => Ok(ContainerState::Exited),
+            "" => Ok(ContainerState::DoesNotExist),
+            _ => eyre::bail!("unknown container state: got {state}"),
+        }
+    }
+
+    #[must_use]
+    pub fn is_stopped(&self) -> bool {
+        matches!(self, Self::Exited | Self::DoesNotExist)
+    }
+
+    #[must_use]
+    pub fn exists(&self) -> bool {
+        !matches!(self, Self::DoesNotExist)
+    }
+}
+
+// it's unlikely that we ever need to erase a line in the destructors,
+// and it's better than keep global state everywhere, or keeping a ref
+// cell which could have already deleted a line
+pub(crate) struct DeleteContainer {
+    engine: Engine,
+    name: String,
+    timeout: u32,
+    color_choice: ColorChoice,
+    verbosity: Verbosity,
+}
+
+pub(crate) fn run_and_get_status(
+    engine: &Engine,
+    args: &[&str],
+    msg_info: &mut MessageInfo,
+) -> Result<ExitStatus> {
+    command(engine)
+        .args(args)
+        .run_and_get_status(msg_info, true)
+}
+
+pub(crate) fn run_and_get_output(
+    engine: &Engine,
+    args: &[&str],
+    msg_info: &mut MessageInfo,
+) -> Result<Output> {
+    command(engine).args(args).run_and_get_output(msg_info)
+}
+
+pub fn container_stop(
+    engine: &Engine,
+    container: &str,
+    timeout: u32,
+    msg_info: &mut MessageInfo,
+) -> Result<ExitStatus> {
+    run_and_get_status(
+        engine,
+        &["stop", container, "--time", &timeout.to_string()],
+        msg_info,
+    )
+}
+
+pub fn container_stop_default(
+    engine: &Engine,
+    container: &str,
+    msg_info: &mut MessageInfo,
+) -> Result<ExitStatus> {
+    // we want a faster timeout, since this might happen in signal
+    // handler. our containers normally clean up pretty fast, it's
+    // only without a pseudo-tty that they don't.
+    container_stop(engine, container, DEFAULT_TIMEOUT, msg_info)
+}
+
+// if stop succeeds without a timeout, this can have a spurious error
+// that is, if the container no longer exists. just silence this.
```

Comment:
another one I missed in the previous pr

this comment feels misplaced

move it to a doccomment or at the call site

Reply:
Since we call `container_rm` from multiple locations, I've updated the comment to better explain why and made it a doc comment:

```rust
/// if stopping a container succeeds without a timeout, this command
/// can fail because the container no longer exists. however, if
/// the container was killed, we need to cleanup the exited container.
/// just silence any warnings.
```